doc/developer/design/20230418_persist_push_based_updates.md

```diff
@@ -0,0 +1,278 @@
+needing to know exactly how, Persist shards are able to update themselves in response to changes made in other processes,
+and it does not need to be tightly bound to the coordination logic elsewhere in Materialize.
+
+Below is a strawman proposal for this interface:
+
+```rust
+#[async_trait]
+pub trait PersistPubSubClient {
+  type Sender: PubSubSender;
+
+  /// Receive handles with which to send requests and receive diffs.
+  async fn connect(
+    addr: String,
+  ) -> Result<(Arc<dyn PubSubSender>, Box<dyn PubSubReceiver>), anyhow::Error>;
+}
+
+#[async_trait]
+pub trait PubSubSender: std::fmt::Debug + Send + Sync {
+  /// Push a diff to subscribers.
+  async fn push(&self, shard_id: &ShardId, diff: &VersionedData) -> Result<(), Error>;
+
+  /// Informs the server which shards this subscribed should receive diffs for.
+  /// May be called at any time to update the set of subscribed shards.
+  async fn subscribe(&self, shards: Vec<ShardId>) -> Result<(), Error>;
+}
+
+pub trait PubSubReceiver: Stream<Item = ProtoPubSubMessage> {}
+```
+
+## Publishing & Applying Updates
+
+When any handle to a persist shard successfully compares-and-sets a new state, it will publish the diff to `PubSubSender`.
+Publishing the diff is not required for correctness in case the process crashes, hits a network partition, or otherwise
+chooses not to publish it (e.g. feature flag).
+
+When a process receives a diff, it needs to apply it to the local copy of state and notify any readers that new data /
+progress may be visible. The mechanisms to do this have largely been built:
+
+* [#18488](https://github.com/MaterializeInc/materialize/pull/18488) gives each process a single, shared copy of persist
+  state per shard. This gives us a natural entrypoint to applying the diff.
+* [#18778](https://github.com/MaterializeInc/materialize/pull/18778) introduces the ability to `await` a change to state,
+  rather relying on polling.
+
+It is not assumed for message delivery to be perfectly reliable nor ordered. We will continue to use polling as a fallback
+for discovering state updates, though with a more generous retry policy than we currently have. If a process receives a
+diff that does not apply cleanly (e.g. out-of-order delivery), the process will fetch the latest state from Cockroach
+directly instead. While we will continue to have these fallbacks, we anticipate message delivery being effective enough
+to eliminate the need for nearly all Cockroach reads.
+
+# Reference Explanation
+
+As a first implementation of the Persist sidechannel, we propose building a new RPC service that flows through `environmentd`.
+The service would be backed by gRPC with a minimal surface area -- to start, we would use a simple bidirection stream that
+allows each end to push updates back and forth, with `environmentd` broadcasting out any changes it receives.
+
+## RPC Service
```

Comment:
This is maybe a bit too cute, but we could also GRPCify the existing `Consensus` interface, and use a "long polling" model for `scan` (where we block for some timeout before returning an empty list of changes, and if a relevant `compare_and_set` comes in in the meantime we resolve any pending `scan`s)  instead of push-based updates.

(I think that description might be too terse, but if it seems intriguing happy to discuss sync.)

Reply:
👍 let's sync up through some medium that is more responsive than GH comments